storage/src/journal/fixed.rs

```diff
@@ -381,79 +384,70 @@ impl<E: Storage + Metrics, A: Codec<Cfg = ()> + FixedSize> Journal<E, A> {
+        let mut sizes = Vec::with_capacity(blobs.len());
+        for b in blobs.iter() {
+            sizes.push(b.1.size().await);
+        }
+        // Create an iterator of (iteration, blob_index, blob, size) tuples to stream.
+        let blob_plus = blobs
+            .into_iter()
+            .zip(sizes.into_iter())
+            .enumerate()
+            .map(|(i, ((blob_index, blob), sz))| (i, *blob_index, (*blob).clone(), sz));
 
-        // Replay all blobs concurrently and stream items as they are read (to avoid occupying too
-        // much memory with buffered data)
         let items_per_blob = self.cfg.items_per_blob;
-        Ok(stream::iter(blobs)
-            .map(move |(index, blob)| async move {
-                // Create buffered reader
-                let size = blob.size().await;
-                let reader = Read::new(blob, size, buffer);
-
-                // Read over the blob in chunks of `CHUNK_SIZE` bytes
-                stream::unfold(
-                    (index, reader, 0u64),
-                    move |(index, mut reader, mut offset)| async move {
-                        // Check if we are at the end of the blob
-                        if offset >= size {
-                            return None;
-                        }
+        let start_offset = (start_pos % items_per_blob) * Self::CHUNK_SIZE_U64;
+
+        let stream = stream::iter(blob_plus).flat_map(move |(i, blob_index, blob, sz)| {
+            // Create a new reader and buffer for each blob.
+            let mut reader = Read::new(blob, sz, buffer);
+            let buf = vec![0u8; Self::CHUNK_SIZE];
+            let initial_offset = if i == 0 {
+                // If this is the very first blob then we need to seek to the starting position.
+                reader.seek_to(start_offset).expect("invalid start_pos");
+                start_offset
+            } else {
+                0
+            };
 
-                        // Seek past any items that precede `start_pos`.
-                        let mut item_pos = items_per_blob * *index + offset / Self::CHUNK_SIZE_U64;
-                        if item_pos < start_pos {
-                            offset += Self::CHUNK_SIZE_U64 * (start_pos - item_pos);
-                            item_pos = start_pos;
-                            if let Err(err) = reader.seek_to(offset) {
-                                return Some((Err(Error::Runtime(err)), (index, reader, offset)));
-                            }
+            stream::unfold(
+                (buf, reader, initial_offset),
+                move |(mut buf, mut reader, offset)| async move {
+                    if offset >= reader.blob_size() {
+                        return None;
+                    }
+                    let item_pos = items_per_blob * blob_index + offset / Self::CHUNK_SIZE_U64;
+                    match reader.read_exact(&mut buf, Self::CHUNK_SIZE).await {
```

Comment:
I don't think we need to clear `buf` here (given the invariants of read_exact) but you may want to note we assume that.

Reply:
done